### DISCUSSION TITLE
发一个openclash与Mosdns和AdguardHome的配置，看看这样子可以吗

### DISCUSSION THREAD

**Discussion by MURChen** [vernesong/OpenClash#2995] [Ideas]
在较新版本的openclash里，Yacd和dashboard面板里无法显示域名，测试发现openclash的Dns劫持里要打开防火墙转发才能显示域名
但分流正常，例如打开油管是能正常从mosdns里得到解析，adguardhome日志里有，openclash的dashboard能显示规则，能正常分流，就是dashboard的域名一栏只有解析后的ip，没有域名
今天在备份好openwrt系统后反复修改这三个的配置，现在找到一个能用的，所以发出来给大家看看有没有问题

##openclash设置
Dns设置
1.dns劫持 防火墙转发
2.自定义dns设置为 127.0.0.1:1745 (adguardhome监听端口 nameserver和fallback均相同）
![image_2023-02-06_16-30-52](https://user-images.githubusercontent.com/47627266/216925335-5553b6a5-a488-4fa7-8a31-a5464cdc859a.jpg)

##mosdns设置
监听端口为 0.0.0.0：6010 （我的mosdns设置参照easymosdns这个项目）
![Snipaste_2023 02 06 16;32;11](https://user-images.githubusercontent.com/47627266/216925457-113ff7c5-ddeb-45b3-bc6b-e67191406566.png)

##AdguardHome设置
1.重定向为 无
2.监听端口为1745
3. dns设置为mosdns的端口 127.0.0.1：6010
 ![Snipaste_2023 02 06 16;31;49](https://user-images.githubusercontent.com/47627266/216925911-3ae3080b-1325-46f7-a108-c3c0ad8b76e2.png)

修改后 我认为是从 openclash 》adguardhome 》mosdns 的顺序
现在adguardhome正常日志，mosdns正常解析，openclash正常分流且dashboard显示域名

**Comment by swirly0p** [comment]
感觉会有的问题：

- AdGuardHome目前看到所有请求应该都来自127.0.0.1
- openclash变成了最后接收IP，再根据IP进行分流

简单讲讲我的理解：

1. dnsmasq转发：
DNS请求→openwrt dnsmasq→openclash（能直接拿到域名）
2. 防火墙转发：
DNS请求→openclash（能直接拿到域名）
3. 我认为更好的选择：
DNS请求→AdGuardHome→openclash（能直接拿到域名）
使用FW4将DNS请求劫持到AdGuardHome，AdGuardHome把openclash作为上游解析器，由openclash负责最终的解析与分流
此时openclash不要开启`本地DNS劫持`，不要使用`绕过中国大陆IP`

**Reply by MURChen** [reply]
我在v2ex上也发了份，现在方案也变了
下面是网友的原话
"1 、重定向 53 到 adg ，关闭缓存，只做过滤，上游填写 openclash 的 dns 服务端口。 
2 、openclash 的 dns 上游填写 mosdns 的端口，关闭缓存，不要劫持。 
3 、设置 mosdns 分流解析，设置缓存策略，国外域名的 dns 记得加入到 openclash 的白名单强制走代理。“”

**Reply by MURChen** [reply]
我开了绕过大陆ip，现在方案蛮稳定的

**Reply by lonecale** [reply]
> 我开了绕过大陆ip，现在方案蛮稳定的

你现在得方案什么样子的，能贴出来了看看吗？你有没有发现mosdns返回的都是ip不是域名

**Reply by MURChen** [reply]
> > 我开了绕过大陆ip，现在方案蛮稳定的
> 
> 你现在得方案什么样子的，能贴出来了看看吗？你有没有发现mosdns返回的都是ip不是域名

没变
1 、重定向 53 到 adg ，关闭缓存，只做过滤，上游填写 openclash 的 dns 服务端口。
2 、openclash 的 dns 上游填写 mosdns 的端口，关闭缓存，不要劫持。
3 、设置 mosdns 分流解析，设置缓存策略，国外域名的 dns 记得加入到 openclash 的白名单强制走代理

**Reply by curionox** [reply]
老哥，第三步mosdns，国外域名的 dns 记得加入到 openclash 的白名单强制走代理 这步要怎么设置？

**Reply by abigchopstick** [reply]
请问你这个意思是不是clash就不用配置文件了，直接用全局代理+绕开大陆IP？
能不能把你的配置文件贴来看一下？包括openclash的

**Reply by MURChen** [reply]
> 请问你这个意思是不是clash就不用配置文件了，直接用全局代理+绕开大陆IP？
> 能不能把你的配置文件贴来看一下？包括openclash的

还是要的，配置文件主要是为了分流，例如一些网站得走不同节点之类的
我自建订阅转换的后端然后自己照着asl等大众的规则来改的
几个月下来，mosdns这个对我来说似乎已经不重要了，感觉dns分流做了没啥明显效果，现在我都不去管它

**Reply by lonecale** [reply]
> > 请问你这个意思是不是clash就不用配置文件了，直接用全局代理+绕开大陆IP？
> > 能不能把你的配置文件贴来看一下？包括openclash的
> 
> 还是要的，配置文件主要是为了分流，例如一些网站得走不同节点之类的
> 我自建订阅转换的后端然后自己照着asl等大众的规则来改的
> 几个月下来，mosdns这个对我来说似乎已经不重要了，感觉dns分流做了没啥明显效果，现在我都不去管它

其实你可以试试 adg上游 mosdns，mosdns海外转发op dns端口 大陆直接请求大陆dns 。然后op dns劫持 停用，只开自定义上游dns 写几个用于海外的dns 这样只有分流海外才走op。然后开启绕过大陆，因为op的防火墙机制 哪怕你dns劫持 停用了，防火墙也会在最后把tcp流量重定向到op，除了几个特殊条件，开启绕过大陆这样的话mosdns大陆ip的就是mosdns内的大陆dns 请求返回客户端 海外的转发op,op解析，一些漏网的也会在最后被op防火墙劫持tcp到op

**Reply by MURChen** [reply]
> > > 请问你这个意思是不是clash就不用配置文件了，直接用全局代理+绕开大陆IP？
> > > 能不能把你的配置文件贴来看一下？包括openclash的
> > 
> > 还是要的，配置文件主要是为了分流，例如一些网站得走不同节点之类的
> > 我自建订阅转换的后端然后自己照着asl等大众的规则来改的
> > 几个月下来，mosdns这个对我来说似乎已经不重要了，感觉dns分流做了没啥明显效果，现在我都不去管它
> 
> 其实你可以试试 adg上游 mosdns，mosdns海外转发op dns端口 大陆直接请求大陆dns 。然后op dns转发停用，只开自定义上游dns 写几个用于海外的dns 这样只有分流海外才走op。然后开启绕过大陆，因为op的防火墙机制 哪怕你停用了，防火墙也会在最后把tcp流量重定向到7892代理端口，除了几个条件，开启绕过大陆这样的话mosdns大陆的就是mosdns内的大陆dns请求返回客户端 海外的转发op,op解析

我跟这个方案差不多，只是把海外dns写在mosdns上用socks5和op连接，国内走阿里dns+op绕过大陆，效果一样

**Reply by lonecale** [reply]
> > > > 请问你这个意思是不是clash就不用配置文件了，直接用全局代理+绕开大陆IP？
> > > > 能不能把你的配置文件贴来看一下？包括openclash的
> > > 
> > > 还是要的，配置文件主要是为了分流，例如一些网站得走不同节点之类的
> > > 我自建订阅转换的后端然后自己照着asl等大众的规则来改的
> > > 几个月下来，mosdns这个对我来说似乎已经不重要了，感觉dns分流做了没啥明显效果，现在我都不去管它
> > 
> > 其实你可以试试 adg上游 mosdns，mosdns海外转发op dns端口 大陆直接请求大陆dns 。然后op dns转发停用，只开自定义上游dns 写几个用于海外的dns 这样只有分流海外才走op。然后开启绕过大陆，因为op的防火墙机制 哪怕你停用了，防火墙也会在最后把tcp流量重定向到7892代理端口，除了几个条件，开启绕过大陆这样的话mosdns大陆的就是mosdns内的大陆dns请求返回客户端 海外的转发op,op解析
> 
> 我跟这个方案差不多，只是把海外dns写在mosdns上用socks5和op连接，国内走阿里dns+op绕过大陆，效果一样


socks5和op连接 这样你可能看到很多dns的链接面板里？其实dnsmasq劫持不清楚为什么除了dnsmasp端口还会劫持53，如果只劫持dnsmasp端口，那理想方案 更改dnsmasp端口 然后adg做53 上游mosdns ,mosdns大陆走大陆dns，海外转发dnsmasp端口,然后op dns劫持dnsmasp 可能更好

**Reply by MURChen** [reply]
> > > > > 请问你这个意思是不是clash就不用配置文件了，直接用全局代理+绕开大陆IP？
> > > > > 能不能把你的配置文件贴来看一下？包括openclash的
> > > > 
> > > > 还是要的，配置文件主要是为了分流，例如一些网站得走不同节点之类的
> > > > 我自建订阅转换的后端然后自己照着asl等大众的规则来改的
> > > > 几个月下来，mosdns这个对我来说似乎已经不重要了，感觉dns分流做了没啥明显效果，现在我都不去管它
> > > 
> > > 其实你可以试试 adg上游 mosdns，mosdns海外转发op dns端口 大陆直接请求大陆dns 。然后op dns转发停用，只开自定义上游dns 写几个用于海外的dns 这样只有分流海外才走op。然后开启绕过大陆，因为op的防火墙机制 哪怕你停用了，防火墙也会在最后把tcp流量重定向到7892代理端口，除了几个条件，开启绕过大陆这样的话mosdns大陆的就是mosdns内的大陆dns请求返回客户端 海外的转发op,op解析
> > 
> > 我跟这个方案差不多，只是把海外dns写在mosdns上用socks5和op连接，国内走阿里dns+op绕过大陆，效果一样
> 
> 
> socks5和op连接 这样你可能看到很多dns的链接面板里？其实dnsmasq劫持不清楚为什么除了dnsmasp端口还会劫持53，如果只劫持dnsmasp端口，那理想方案 更改dnsmasp端口 然后adg做53 上游mosdns ,mosdns大陆走大陆dns，海外转发dnsmasp端口,然后op dns劫持dnsmasp 可能更好
> 
> 

并没有，面板就两条连接，op没开劫持，dns先到adg，adg到op，op写mosdns，所以说我的方案一个挂全挂

**Comment by GD2021** [comment]
能把mosdns的配置贴出来嘛？

**Comment by GeekXtop** [comment]
> 我在v2ex上也发了份，现在方案也变了 下面是网友的原话 "1 、重定向 53 到 adg ，关闭缓存，只做过滤，上游填写 openclash 的 dns 服务端口。 2 、openclash 的 dns 上游填写 mosdns 的端口，关闭缓存，不要劫持。 3 、设置 mosdns 分流解析，设置缓存策略，国外域名的 dns 记得加入到 openclash 的白名单强制走代理。“”

大佬，我也是这样配置的，我没用adg，mosdns设置dnsmasq转发，openclash不劫持，监听mosdns的5335端口，但是一直报WARN	forward_local	upstream error context deadline exceeded

**Reply by immaiker911** [reply]
我也是这个问题，一直没得到好的解决方案，不知道是啥原因，你这边解决了吗

**Comment by MURChen** [comment]
mosdns现在对我来说只是作为一个好用的dns缓存器，没用mosdns前我单独使用adguardhome，那时它的lazy cache有点问题且那时候adg与openclash配合不好，本来想用mosdns让dns解析与openclash解耦合，但现在我的方案反而更加依赖3个应用，任何一个挂了都会上不了网，所以有能力想玩的不用看这个方案了，现在我也没想法去整这些，最多更新更新自己写的分流规则

**Reply by lonecale** [reply]
Adg有个后备dns 写一个其他的这样只要adg不坏，其他的坏了不至于断网走后备dns

**Reply by MURChen** [reply]
好像九月份添加了这功能，b37添加的，最近没去看这些，的确这功能可以的，等会试试op断了能不能转过去

**Reply by MURChen** [reply]
嗯，这功能可以，这样只要adg不挂，mosdns&op挂了dns也是正常的
